File: Frameworks/react-web/src/models/ods-app.ts
import { parseAuth, type OdsAuth } from './ods-auth.ts'
import { parseTheme, type OdsTheme } from './ods-theme.ts'
import { parseAppSetting, type OdsAppSetting } from './ods-app-setting.ts'
import { parseDataSource, type OdsDataSource } from './ods-data-source.ts'
import { parseHelp, parseTourStep, type OdsHelp, type OdsTourStep } from './ods-help.ts'
import { parseMenuItem, type OdsMenuItem } from './ods-menu-item.ts'
import { parsePage, type OdsPage } from './ods-page.ts'

/** The top-level model representing a complete ODS application. */
export interface OdsApp {
  appName: string
  /** Optional emoji or icon identifier for the app (top-level identity). */
  appIcon?: string
  /** Logo URL shown in sidebar/drawer header. */
  logo?: string
  /** Favicon URL shown in browser tab. */
  favicon?: string
  startPage: string
  /** Role-based start pages (e.g. { admin: 'dashPage', manager: 'reportPage' }). */
  startPageByRole: Record<string, string>
  menu: OdsMenuItem[]
  pages: Record<string, OdsPage>
  dataSources: Record<string, OdsDataSource>
  help?: OdsHelp
  tour: OdsTourStep[]
  settings: Record<string, OdsAppSetting>
  auth: OdsAuth
  theme: OdsTheme
}

/** Resolve startPage to a string. Accepts "pageName" or { default: "pageName", admin: "other" }. */
function parseStartPage(raw: unknown): string {
  if (typeof raw === 'string') return raw
  if (raw && typeof raw === 'object' && 'default' in raw) return (raw as Record<string, string>).default
  return ''
}

/** Build the role→page map, excluding the `default` key (stored in startPage). */
function parseStartPageByRole(raw: unknown): Record<string, string> {
  if (typeof raw === 'string') return {}
  if (raw && typeof raw === 'object') {
    return Object.fromEntries(
      Object.entries(raw as Record<string, unknown>)
        .filter(([k, v]) => k !== 'default' && typeof v === 'string')
        .map(([k, v]) => [k, v as string])
    )
  }
  return {}
}

export function parseApp(json: unknown): OdsApp {
  const j = json as Record<string, unknown>
  const pagesRaw = j['pages'] as Record<string, unknown> | undefined
  const dsRaw = j['dataSources'] as Record<string, unknown> | undefined
  const settingsRaw = j['settings'] as Record<string, unknown> | undefined

  return {
    appName: j['appName'] as string,
    appIcon: j['appIcon'] as string | undefined,
    logo: j['logo'] as string | undefined,
    favicon: j['favicon'] as string | undefined,
    startPage: parseStartPage(j['startPage']),
    startPageByRole: parseStartPageByRole(j['startPage']),
    menu: Array.isArray(j['menu'])
      ? (j['menu'] as unknown[]).map(parseMenuItem)
      : [],
    pages: pagesRaw
      ? Object.fromEntries(
          Object.entries(pagesRaw).map(([k, v]) => [k, parsePage(v)])
        )
      : {},
    dataSources: dsRaw
      ? Object.fromEntries(
          Object.entries(dsRaw).map(([k, v]) => [k, parseDataSource(v)])
        )
      : {},
    help: parseHelp(j['help']),
    tour: Array.isArray(j['tour'])
      ? (j['tour'] as unknown[]).map(parseTourStep)
      : [],
    settings: settingsRaw
      ? Object.fromEntries(
          Object.entries(settingsRaw).map(([k, v]) => [k, parseAppSetting(v)])
        )
      : {},
    auth: parseAuth(j['auth']),
    theme: parseTheme(j['theme']),
  }
}

File: Frameworks/flutter-local/lib/models/ods_app.dart
import 'ods_app_setting.dart';
import 'ods_auth.dart';
import 'ods_data_source.dart';
import 'ods_help.dart';
import 'ods_menu_item.dart';
import 'ods_page.dart';
import 'ods_theme.dart';

/// The top-level model representing a complete ODS application.
///
/// ODS Spec alignment: This is the root object of ods-schema.json. It holds
/// everything the framework needs to render a full application: metadata
/// (appName), navigation (startPage, menu), content (pages), data layer
/// (dataSources), and user guidance (help, tour).
///
/// ODS Ethos: One JSON file = one complete application. No external config,
/// no build steps, no server. This model captures that entire contract.
///
/// Architecture note: This class lives in the models layer and has ZERO
/// Flutter imports. It is pure Dart, making it reusable by future non-Flutter
/// ODS frameworks and by the planned code-generation off-ramp.
class OdsApp {
  final String appName;

  /// Optional emoji or icon identifier for the app (top-level identity).
  final String? appIcon;

  /// Logo URL shown in sidebar/drawer header.
  final String? logo;

  /// Favicon URL shown in browser tab (web only; ignored on Flutter desktop).
  final String? favicon;

  /// The default start page ID (used when no role-specific page matches).
  final String startPage;

  /// Role-based start page overrides. When the spec's `startPage` is an
  /// object, each key is a role and the value is the page ID for that role.
  /// The `default` key is stored in [startPage].
  final Map<String, String> startPageByRole;

  /// Navigation menu entries shown in the drawer.
  /// Optional: apps with a single page may omit the menu.
  final List<OdsMenuItem> menu;

  /// All pages in the app, keyed by page ID.
  final Map<String, OdsPage> pages;

  /// All data sources, keyed by a unique identifier referenced by
  /// list components and submit actions.
  final Map<String, OdsDataSource> dataSources;

  /// Optional in-app help content (overview + per-page contextual help).
  final OdsHelp? help;

  /// Optional guided tour steps shown on first launch.
  final List<OdsTourStep> tour;

  /// Optional user-configurable settings with default values.
  final Map<String, OdsAppSetting> settings;

  /// Authentication and role-based access control configuration.
  /// When absent or `multiUser: false`, the app runs in single-user mode.
  final OdsAuth auth;

  /// Visual theme + customizations.
  final OdsTheme theme;

  /// Resolves the start page for the given roles. Returns the first
  /// role-specific match, or falls back to [startPage].
  String startPageForRoles(List<String> roles) {
    for (final role in roles) {
      final page = startPageByRole[role];
      if (page != null) return page;
    }
    return startPage;
  }

  const OdsApp({
    required this.appName,
    this.appIcon,
    this.logo,
    this.favicon,
    required this.startPage,
    this.startPageByRole = const {},
    required this.menu,
    required this.pages,
    required this.dataSources,
    this.help,
    this.tour = const [],
    this.settings = const {},
    this.auth = const OdsAuth(),
    this.theme = const OdsTheme(),
  });

  factory OdsApp.fromJson(Map<String, dynamic> json) {
    // startPage can be a string or an object with role-based mappings
    final rawStartPage = json['startPage'];
    final String startPage;
    final Map<String, String> startPageByRole;
    if (rawStartPage is Map) {
      final map = Map<String, String>.from(rawStartPage.map(
        (k, v) => MapEntry(k as String, v as String),
      ));
      startPage = map.remove('default') ?? map.values.first;
      startPageByRole = map;
    } else {
      startPage = rawStartPage as String;
      startPageByRole = const {};
    }

    return OdsApp(
      appName: json['appName'] as String,
      appIcon: json['appIcon'] as String?,
      logo: json['logo'] as String?,
      favicon: json['favicon'] as String?,
      startPage: startPage,
      startPageByRole: startPageByRole,
      // Menu is optional — default to empty list if not provided.
      menu: (json['menu'] as List<dynamic>?)
              ?.map((m) => OdsMenuItem.fromJson(m as Map<String, dynamic>))
              .toList() ??
          [],
      pages: (json['pages'] as Map<String, dynamic>).map(
        (key, value) => MapEntry(key, OdsPage.fromJson(value as Map<String, dynamic>)),
      ),
      // DataSources are optional — apps with only static text don't need them.
      dataSources: (json['dataSources'] as Map<String, dynamic>?)?.map(
            (key, value) =>
                MapEntry(key, OdsDataSource.fromJson(value as Map<String, dynamic>)),
          ) ??
          {},
      help: json['help'] != null
          ? OdsHelp.fromJson(json['help'] as Map<String, dynamic>)
          : null,
      tour: (json['tour'] as List<dynamic>?)
              ?.map((t) => OdsTourStep.fromJson(t as Map<String, dynamic>))
              .toList() ??
          [],
      settings: (json['settings'] as Map<String, dynamic>?)?.map(
            (key, value) =>
                MapEntry(key, OdsAppSetting.fromJson(value as Map<String, dynamic>)),
          ) ??
          {},
      auth: OdsAuth.fromJson(json['auth'] as Map<String, dynamic>?),
      theme: OdsTheme.fromJson(json['theme'] as Map<String, dynamic>?),
    );
  }
}
